### PR TITLE
Added achievements!

### LINKED ISSUE #12: Add Achievements
I would love to Challenge myself by collecting achievements which I can brag about in my friend circle ;)
Achievements like : 

- Reveals all the cells in Beginner Mode within 100 seconds . 
- Reveals all the cells in Intermediate Mode within 100 seconds . 
- Reveals all the cells in Expert Mode within 100 seconds . 

- Reveals all the cells in Beginner Mode within 50 seconds . 
- Reveals all the cells in Intermediate Mode within 50 seconds . 
- Reveals all the cells in Expert Mode within 50 seconds . 

-Revealing all the cells for the first time. 
-Revealing all the cells for the 5th time.

### PR DESCRIPTION
**Pull request by R4H33M**
Fixes #12 

image

### CODE REVIEW

**Comment by Begasus** [comment]
LGTM

File: src/AchievementWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+}
+
+void
+AchievementWindow::UpdateLabels(void)
+{
+	for (int x = 0; x < 4; x++) {
+		for (int y = 0; y < 2; y++) {
+			if (gAchievements[x][y]) {
+				labelList[x][y]->SetHighColor(1,50,32);
```
Comment (humdingerb) [inline src/AchievementWindow.cpp]: It'd be nicer to use different colours to make it easier to see which achievements are unlocked. For example:

```
Unlocked: labelList[x][y]->SetHighColor(0,150,32);
Locked: labelList[x][y]->SetHighColor(80,80,80);
```
Not necessarily these colours, pick what you like...

**Comment by R4H33M** [comment]
Good colour suggestions. I've added them. Looks much better now!

**Comment by humdingerb** [comment]
Alright! Before I merge, is there a reason to leave the string translation as a TODO?
You could just quickly add the B_TRANSLATE macros to all GUI strings and be done with it...

**Comment by R4H33M** [comment]
Is this what you were thinking? I am not really sure how Haiku's localisation system works, but does it look right?

File: src/AchievementWindow.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
+#include "AchievementWindow.h"
+#include <Button.h>
+#include <Catalog.h>
+#include "Globals.h"
+#include <LayoutBuilder.h>
+
+#define M_RESET_ACHIEVEMENTS 'rsac'
+
+#define B_TRANSLATION_CONTEXT "AchievementWindow"
```
Comment (humdingerb) [inline src/AchievementWindow.cpp:11]: Add before this:

`#undef B_TRANSLATION_CONTEXT`

File: src/AchievementWindow.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+AchievementWindow::AchievementWindow(void)
+	:	DWindow(BRect(100,500,100,400),B_TRANSLATE("Achievements"),
+		B_TITLED_WINDOW, B_NOT_RESIZABLE |B_NOT_ZOOMABLE |
+		B_AUTO_UPDATE_SIZE_LIMITS)
+
+{
+	BView *top = GetBackgroundView();
+
+	BStringView *begTitle, *intTitle, *expTitle, *genTitle;
+	BButton *close = new BButton("ok", "OK", new BMessage(B_QUIT_REQUESTED));
```
Comment (humdingerb) [inline src/AchievementWindow.cpp]: B_TRANSLATE("OK")

File: src/AchievementWindow.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	BButton *reset = new BButton("reset", "Reset", new BMessage(M_RESET_ACHIEVEMENTS));
```
Comment (humdingerb) [inline src/AchievementWindow.cpp]: B_TRANSLATE("Reset")

**Review by humdingerb — CHANGES_REQUESTED** [review]
You update the locales/en.catkeys with `make catkeys` after a successful `make`.
I can then upload it to https://i18n.kacperkasper.pl/projects and the translators get to work on that.
Please include the new en.catkeys in your PR.

**Comment by R4H33M** [comment]
Ahhh that's how it works. Cool! 
All done!

**Comment by humdingerb** [comment]
It's slightly awkward if I unlocked an achievment and at the same time got a new high score. In that case, I get the higscore window, but the achievement alert floats on top of it.

How about not showing the achivements alert but directly the achivements window with the title "New Achievments unlocked!"?

**Comment by R4H33M** [comment]
So completely remove the achievements alert, and instead just show the achievements window?

**Comment by humdingerb** [comment]
> So completely remove the achievements alert, and instead just show the achievements window?

Yes. I'd say that'd be an easy cheat/work-around. Make sure to first pop up the Achievements window, then the New-Highscore, so the latter has focus and the user can enter their name there.

The Achievments window's title should be "New Achievement unlocked!".

**Comment by R4H33M** [comment]
Is this what you were thinking?

File: src/MainWindow.cpp [hunk 5]
```diff
@@ -425,20 +427,13 @@ MainWindow::AchievementCheck(void)
 		numAchieved++;
 	}
 
-	BAlert *achievementAlert;
-
+	AchievementWindow *achievementwin = new AchievementWindow();
+
```
Comment (humdingerb) [inline src/MainWindow.cpp:429]: stray tab.

File: src/MainWindow.cpp [hunk 6]
```diff
@@ -425,20 +427,13 @@ MainWindow::AchievementCheck(void)
 		numAchieved++;
 	}
 
-	BAlert *achievementAlert;
-
+	AchievementWindow *achievementwin = new AchievementWindow();
+	
 	if (numAchieved > 1) {
```
Comment (humdingerb) [inline src/MainWindow.cpp]: The below does work of course, but if you want to learn now to perfectly do this, that will also guarantee correct localization in any language, study this, for example:
https://git.haiku-os.org/haiku/tree/src/preferences/filetypes/ApplicationTypesWindow.cpp#n349

The title should of course also be B_TRANSLATE()d.

Comment (R4H33M) [reply src/MainWindow.cpp]: I see - thanks for pointing it out.

**Comment by R4H33M** [comment]
Didn't know about BStringFormat. Cool!
Ready for review!

File: src/MainWindow.cpp [hunk 7]
```diff
@@ -1,4 +1,5 @@
 #include <AboutWindow.h>
+#include <Alert.h>
```
Comment (humdingerb) [inline src/MainWindow.cpp]: not needed anymore

File: src/AchievementWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
+#include "AchievementWindow.h"
```
Comment (humdingerb) [inline src/AchievementWindow.cpp:1]: It's nice to have the system headers and app headers grouped together (alpha-sorted).

File: src/AchievementWindow.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+#include <Button.h>
+#include <Catalog.h>
+#include "Globals.h"
+#include <LayoutBuilder.h>
+
+#undef B_TRANSLATION_CONTEXT
```
Comment (humdingerb) [inline src/AchievementWindow.cpp]: Maybe move in front of the #define B_TRANSLATION_CONTEXT

File: src/MainWindow.cpp [hunk 10]
```diff
@@ -382,6 +395,50 @@ MainWindow::HighScoreCheck(void)
...
+		gAchievements[gDifficulty][1] = true;
+		numAchieved++;
+	}
+
+	AchievementWindow *achievementwin = new AchievementWindow();
+	BString title;
+	static BStringFormat format(B_TRANSLATE("{0, plural,"
+		"=1{New achievement unlocked!}"
+		"other{New achievements unlocked!}}"));
+	format.Format(title, numAchieved);
```
Comment (humdingerb) [inline src/MainWindow.cpp]: JFYI, if you don't already know: you can even use "#" to get the numAchieved inserted into the string, like

`other{# new achievements unlocked!}`

Though I see why you may want to keep the window title a bit shorter...

File: src/MainWindow.cpp [hunk 11]
```diff
@@ -382,6 +395,50 @@ MainWindow::HighScoreCheck(void)
...
+	achievementwin->SetTitle(title);
+	
+	if (numAchieved > 0)
+		achievementwin->Show();
```
Comment (humdingerb) [inline src/MainWindow.cpp:438]: You can move the whole window creation etc. into this if clause.

**Review by humdingerb — CHANGES_REQUESTED** [review]
Only a few more details and we're done. :)
Don't forget to re-create the en.catkeys.

File: src/MainWindow.cpp (hunk 10 above)
Comment (R4H33M) [reply src/MainWindow.cpp]: Yeah. it's a little on the long side, so I suggest we keep it how it is.

**Comment by R4H33M** [comment]
Done!

**Comment by humdingerb** [comment]
Thanks for keeping at it all this time... :)

**Pull request merged**